### ISSUE TITLE
Notes about Hydra

### ISSUE THREAD

**Issue by andrewjong** [kourai-research/pytorch-boilerplate#4]
Overriding with user configs isn't the easiest apparently. There is a kind of hacky work around

See https://github.com/facebookresearch/hydra/issues/386#issuecomment-589586845

**Comment by andrewjong** [comment]
I think Hydra actually isn't the best for our experiment-driven use-case. Since the above showed that user configs aren't easy, the only advantage of Hydra that I see is organization; but I don't like having the args separated from the code. It's too redundant imo.

I much prefer putting the args within the class definition itself, like in PT Lightning. That way when you're reading the code, you immediately see what options it provides.
https://pytorch-lightning.readthedocs.io/en/latest/hyperparameters.html#argparser-best-practices

In terms of hierarchical organization, it's easy to just prepend a ".class_name" before the arg. For example, "models.ResNet.num_layers". These dots can be dynamically added based on the path relative to the project root.

Then we can have a `python train.py --make-config-template desired/output/of/template` which generates a hierarchical .yaml file based on the dots of the args. The default values can have a `#default` appended to the side. This also has the advantage of having a single file to edit for all the args.

Then we do `python train.py --config the/config/i/modified`. It would then be quite easy to load in the config from the hierarchical .yaml.

If we want to make a config template by copying a previous one, we can do
`python train.py --make-config-template desired/output/of/template --template-from template/to/copy/from`. By default, --template-from would be set to None and use the default args from ArgParse.

**Comment by SBoulanger** [comment]
I like it. Way simpler, less clutter file-wise, nothing new for people to learn.

I like the addition of templates. It would give the ability to maintain a working config as well. Which could be a preference to flags on the CLI.

**Comment by SBoulanger** [comment]
I am second guessing our position on this. Check out the file linked below. The configs get messy really quick. And when we are modularizing the repo, I am having a hard time keeping it organized.

https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/domain_templates/imagenet.py

**Comment by SBoulanger** [comment]
I tried hydra and dont like it either. Lets just get this part right. 

- [ ] Clean
- [ ] Extendable

**Comment by andrewjong** [comment]
What about the PT Lightning config was messy? I feel like if they just pulled the args to the top of the class, then moved the global args to another file, it'd be more organized.

**Comment by SBoulanger** [comment]
So for every custom class, we would require a static method that adds all the args they would need?

**Comment by SBoulanger** [comment]
If you look at the mnist example. You'll see what I am talking about in the train.py file.

**Comment by andrewjong** [comment]
Where's the MNIST example? Can't find it on Google

>So for every custom class, we would require a static method that adds all the args they would need?
Yeah. The way the class is configged is present in the class itself. It's like a more detailed constructor.

Could be familiarity bias on my part. This makes sense to me because it's what was used in the [Pix2Pix/CycleGAN repository](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/cb2c4f89de5a2ef454d146c1aed97300d59abc7a/models/cycle_gan_model.py#L20)

**Comment by SBoulanger** [comment]
> Where's the MNIST example? Can't find it on Google
https://github.com/kourai-research/mnist

**Comment by SBoulanger** [comment]
I don't know any other very good way of handing the parameters honestly

**Comment by andrewjong** [comment]
Ohh yeah that is kind of annoying. Maybe if it's done in a separate function it'll look nicer. [This](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/cb2c4f89de5a2ef454d146c1aed97300d59abc7a/options/base_options.py#L60) is what they do in CycleGAN.

Could also add the parameters dynamically actually. Do a for loop that sweeps all Python files and finds any subclasses of the Lightning module, then read in those params.

**Comment by SBoulanger** [comment]
Yeah that is definitely a way to make it nicer. I also would want to make defaults as class attributes. I don't like how unclear the defaults are when looking at a class.

**Comment by SBoulanger** [comment]
Class initialization is another thing we should consider. Do we want every class to take in hparams? Which should and which shouldnt?

**Comment by andrewjong** [comment]
I think project-specific things should take in hparams, while project-independent things should be initialized with their own parameters.

So e.g. Systems take in hparams because they're part of the project. But nn.Modules should take args so that they can be plug-and-play in other projects.

**Comment by SBoulanger** [comment]
I like that.

**Comment by omry** [comment]
There is a reason this feature request is there, this is one of the current limitations and I am absolutely planning to address this, most likely in Hydra 1.1.
1.0 is just around the corner and is already adding a ton of new functionality and I don't want to delay it further.

You are of course free in invent you own system, but keep in mind that by using Hydra you will benefit from many things that may not seem apparent yet.

As examples, the next version brings:
* Strongly typed configs ([Structured configs](https://hydra.cc/docs/next/tutorials/structured_config/intro/))
* [Ax ](https://hydra.cc/docs/next/plugins/ax_sweeper/) and [Nevergrad](https://hydra.cc/docs/next/plugins/nevergrad_sweeper/) sweeper plugins for command line driven hyper parameter optimization
* Ray launcher plugin for AWS launching from the command line (in progress)
* [Joblib launcher plugin](https://hydra.cc/docs/next/plugins/joblib_launcher/) for local parallel execution.

In addition, Hydra is already offering some features not covered by your approach like dynamic tab completion for your config, automatic working directory management and automatic textual logging configuration.

Finally, Hydra is getting adopted by some of the largest and most successful machine learning frameworks from Facebook.
[mmf](https://github.com/facebookresearch/mmf) has also switched to using OmegaConf for configuration and will possibly switch to Hydra once some additional support for it's use cases is added.

Other frameworks are in the process (I don't want to name anything not yet public).

Basically, the approach you are proposing here have failed more than one large framework (it is very limiting and does not scale well).
Some such frameworks are already in the process of switching to Hydra.

**Comment by SBoulanger** [comment]
@omry Thanks for letting us know that you guys have this feature request queued. All the other features you mentioned seem great as well.

We'll reconsider our current sub-par approach and I'll take a close look at hydra.

**Comment by andrewjong** [comment]
Hey @omry , thanks for stopping by to comment. Looking forward to Hydra 1.1!
